File: activeobject/src/activation_queue.rb
class ActivationQueue
  def initialize
    @mutex = Mutex.new
    @condition = ConditionVariable.new
    @items = []
  end
  def enqueue(item)
    @mutex.synchronize {
      @items << item
      @condition.signal
    }
  end
  def dequeue
    @mutex.synchronize {
        while @items.empty?
          @condition.wait(@mutex)
        end
        return @items.shift
      }
  end
end

File: activeobject/java/src/activeobject/ActivationQueue.java
package activeobject;

import java.util.ArrayDeque;
import java.util.Deque;

// In reality, we would use something like java.util.concurrent.ArrayBlockingQueue
public class ActivationQueue<T> {
	private Deque<T> items = new ArrayDeque<T>();
	
	public synchronized void enqueue(T item) {
		items.addLast(item);
		notifyAll();
	}
	
	public synchronized T dequeue() {
		while (items.isEmpty()) {
			waitForNotification();
		}
		return items.removeFirst();
	}

	private void waitForNotification() {
		try {
			wait();
		} catch (InterruptedException e) {
			throw new RuntimeException(e);
		}
	}
}
